Fixing up docs

petemoore opened the pull request:
I've no idea if this is going to work at the end, and we'll be able to build docker worker machine images, but since we didn't archive this repo (yet) it can't harm to apply fixes if we spot mistakes.

At the end of the process we can decide if we archive this repo or not, once we know if it provides a reliable way to build the docker-worker machine images or not.

petemoore commented:
CC @escapewindow

jwhitlock commented:
I hit a dead-end - Virtualbox doesn't run on M1 macs. I have my old laptop, I can try there. I suspect a Virtualbox-workflow won't be viable for long.

When I ran `brew install --cask virtualbox`, I got 4 macOS popups about blocked system extensions that the author (Oracle) needs to update. I suspect these are Intel-only extensions.

When I ran `vagrant vbguest`, I got the message:

> A Vagrant environment or target machine is required to run this
> command. Run `vagrant init` to create a new Vagrant environment. Or,
> get an ID of a target machine from `vagrant global-status` to run
> this command on. A final option is to change to a directory with a
> Vagrantfile and to try again.

I think those commands need to be run in the folder with the `Vagrantfile`, or it may be part of my non-working M1 setup.

This command checks out an archived repo:

```
$ git clone git@github.com:taskcluster/docker-worker-deploy.git
```

I think instead it needs to be inside a Taskcluster monorepo, with the ``docker-worker-deploy`` folder at the same level as ``workers/docker-worker``:

```
$ mkdir -p ~/git
$ cd ~/git
$ git clone git@github.com:taskcluster/taskcluster.git
$ cd ~/git/taskcluster/workers
$ git clone git@github.com:taskcluster/docker-worker-deploy.git
$ cd ~/git/taskcluster/workers/docker-worker
$ vagrant up
```

I also tried a simlink to ``workers/docker-worker``, like this:

```
$ mkdir -p ~/git
$ cd ~/git
$ git clone git@github.com:taskcluster/taskcluster.git
$ ln -sf ~/git/taskcluster/workers/docker-worker ~/git/docker-worker
$ git clone git@github.com:taskcluster/docker-worker-deploy.git
$ cd ~/git/docker-worker
$ vagrant up
```

but that failed with ``vm: The host path of the shared folder is missing: ../docker-worker-deploy``, so Virtualbox can't follow symlinks on my system.

This also implies that the `docker-worker-deploy` project belongs in the monorepo. I'd also be interested in trying this with current tools (like docker-compose and [multi-platform docker builds](https://www.docker.com/blog/multi-platform-docker-builds/), which are supported on arm64 / M1 Macs.

petemoore commented:
Thanks for the thorough review @jwhitlock and also for taking the docs for a test-run!

> I hit a dead-end - Virtualbox doesn't run on M1 macs. I have my old laptop, I can try there. I suspect a Virtualbox-workflow won't be viable for long.

Ah interesting. Indeed, vagrant seems to have lost popularity in recent years. Perhaps an alternative technology is a better long-term bet.

> When I ran `brew install --cask virtualbox`, I got 4 macOS popups about blocked system extensions that the author (Oracle) needs to update. I suspect these are Intel-only extensions.

That seems like a reasonable hypothesis, but indeed also adds weight to the argument that vagrant isn't very actively maintained.

> When I ran `vagrant vbguest`, I got the message:
> 
> > A Vagrant environment or target machine is required to run this
> > command. Run `vagrant init` to create a new Vagrant environment. Or,
> > get an ID of a target machine from `vagrant global-status` to run
> > this command on. A final option is to change to a directory with a
> > Vagrantfile and to try again.
> 
> I think those commands need to be run in the folder with the `Vagrantfile`, or it may be part of my non-working M1 setup.

I think we don't have an actual `Vagrantfile` in the repo - I don't know if that is because when this part was split out from docker-worker, it got lost or stayed in the docker-worker repo, or if it is intentional. In any case, I've changed it to `vagrant init` since we don't have the `Vagrantfile`. Thanks for raising!

> This command checks out an archived repo:
> 
> ```
> $ git clone git@github.com:taskcluster/docker-worker-deploy.git
> ```

That is strange, that is this repo that this PR is in. From the github interface, I don't see anything showing that it is archived, and I am able to create a PR against it, so I wonder what is going on there.

> I think instead it needs to be inside a Taskcluster monorepo, with the `docker-worker-deploy` folder at the same level as `workers/docker-worker`:
> 
> ```
> $ mkdir -p ~/git
> $ cd ~/git
> $ git clone git@github.com:taskcluster/taskcluster.git
> $ cd ~/git/taskcluster/workers
> $ git clone git@github.com:taskcluster/docker-worker-deploy.git
> $ cd ~/git/taskcluster/workers/docker-worker
> $ vagrant up
> ```
> 
> I also tried a symlink to `workers/docker-worker`, like this:
> 
> ```
> $ mkdir -p ~/git
> $ cd ~/git
> $ git clone git@github.com:taskcluster/taskcluster.git
> $ ln -sf ~/git/taskcluster/workers/docker-worker ~/git/docker-worker
> $ git clone git@github.com:taskcluster/docker-worker-deploy.git
> $ cd ~/git/docker-worker
> $ vagrant up
> ```
> 
> but that failed with `vm: The host path of the shared folder is missing: ../docker-worker-deploy`, so Virtualbox can't follow symlinks on my system.
> 
> This also implies that the `docker-worker-deploy` project belongs in the monorepo. I'd also be interested in trying this with current tools (like docker-compose and [multi-platform docker builds](https://www.docker.com/blog/multi-platform-docker-builds/), which are supported on arm64 / M1 Macs.

My preference for the desired end state is that we add a single directory [here](https://github.com/mozilla/community-tc-config/tree/main/imagesets) for docker-worker. This would onboard it to the existing image building process for our other workers. My aim with this PR is mostly to get the old system working again (since I believe we can't build docker-worker in automation at the moment) and as soon as we've established all the required steps, we should be able to migrate it over. This PR is mostly a means for me to document corrections while I work on fixing it up, and hopefully we can archive this repo and the monopacker repo etc once we've managed to do that. If it turns out to be too difficult and take too long, we may have to look at the currently mid-to-long-term solutions of replacing docker-worker workers with generic-worker, or creating new docker-worker image building process from scratch.

jwhitlock commented:
> I think we don't have an actual Vagrantfile in the repo - I don't know if that is because when this part was split out from docker-worker, it got lost or stayed in the docker-worker repo, or if it is intentional. In any case, I've changed it to vagrant init since we don't have the Vagrantfile. Thanks for raising!

I was thinking that this `Vagrantfile` was the one that we wanted:

https://github.com/taskcluster/taskcluster/blob/main/workers/docker-worker/Vagrantfile

but maybe a temporary one is OK for getting the tools installed.

> That is strange, that is this repo that this PR is in. From the github interface, I don't see anything showing that it is archived, and I am able to create a PR against it, so I wonder what is going on there.

Sorry for the confusion, I reported the wrong line! This is the one that checks out the archived repo:

```
git clone git@github.com:taskcluster/docker-worker.git
```

I see you've address that in your latest commit.

I have not tried the commands on my old laptop yet, but when I do, I'll try the new instructions.

petemoore commented:
> > I think we don't have an actual Vagrantfile in the repo - I don't know if that is because when this part was split out from docker-worker, it got lost or stayed in the docker-worker repo, or if it is intentional. In any case, I've changed it to vagrant init since we don't have the Vagrantfile. Thanks for raising!
> 
> I was thinking that this `Vagrantfile` was the one that we wanted:
> 
> https://github.com/taskcluster/taskcluster/blob/main/workers/docker-worker/Vagrantfile
> 
> but maybe a temporary one is OK for getting the tools installed.

Perfect, nice spot! I've removed the `vagrant init` and instead moved the vagrant commands later down until the user is already in the docker-worker directory of the monorepo.

> > That is strange, that is this repo that this PR is in. From the github interface, I don't see anything showing that it is archived, and I am able to create a PR against it, so I wonder what is going on there.
> 
> Sorry for the confusion, I reported the wrong line! This is the one that checks out the archived repo:
> 
> ```
> git clone git@github.com:taskcluster/docker-worker.git
> ```
> 
> I see you've address that in your latest commit.

No worries! That makes sense now. :-)

> I have not tried the commands on my old laptop yet, but when I do, I'll try the new instructions.

Thanks John, much appreciated!

I'll see if I can try them out on some fresh machine too.

jwhitlock reviewed (approved):
I'm not going to be able to give a thorough review, and I apologize for not recognizing this earlier. I believe this document update is an improvement, and a good basis for making tweaks in later PRs. Maybe we'll even get to deprecate docker-worker before the next cert update is needed!
